### PR TITLE
ARROW-6177: [C++] Add Array::Validate()

### PR DESCRIPTION
**Pull request by pitrou**
This is in line with ChunkedArray::Validate() and Table::Validate().

### CODE REVIEW

**Review by jorisvandenbossche — APPROVED** [review]

File: cpp/src/arrow/array.h [hunk 1]
```diff
@@ -343,6 +343,15 @@ class ARROW_EXPORT Array {
   /// \return PrettyPrint representation of array suitable for debugging
   std::string ToString() const;
 
+  /// \brief Perform any validation checks to determine obvious inconsistencies
+  /// within the array's internal data.
+  ///
+  /// This can be an expensive check, potentially O(length).
+  ///
+  /// \return Status
+  ARROW_EXPORT
```
Comment (kou) [inline cpp/src/arrow/array.h]: This is needless.

**Review by kou — COMMENTED** [review]
Can we deprecate `arrow::ValidateArray()`?

File: cpp/src/arrow/array.h (hunk 1 above)
Comment (pitrou) [reply cpp/src/arrow/array.h]: Hmm, right.

**Comment by pitrou** [comment]
> Can we deprecate arrow::ValidateArray()?

Why not, though I think keeping it wouldn't harm either.

**Comment by wesm** [comment]
+1 on deprecation

**Review by kou — APPROVED** [review]
+1
I'll merge this.

**Comment by kou** [comment]
Oh, this causes CI failure because https://github.com/apache/arrow/commit/cba9c7b3945f2207d74d33ad407b35e6b362d166 uses `ValidateArray()`.
I've created https://github.com/apache/arrow/pull/5079 to fix this.

**Pull request closed without merging**